src/test/scala/org/apache/spark/sql/delta/OptimisticTransactionSuite.scala

```diff
@@ -34,76 +37,632 @@ class OptimisticTransactionSuite extends QueryTest with SharedSparkSession {
     }
   }
 
   test("allow read+append+delete against no data change") {
     withTempDir { tempDir =>
       val log = DeltaLog(spark, new Path(tempDir.getCanonicalPath))
-      // Initialize the log and add data. Truncate() is just a no-op placeholder.
-      log.startTransaction().commit(addA :: Nil, Truncate())
+      // Initialize the log and add data. ManualUpdate is just a no-op placeholder.
+      log.startTransaction().commit(addA :: Nil, ManualUpdate)
 
       val txn = log.startTransaction()
       txn.filterFiles()
       val winningTxn = log.startTransaction()
-      winningTxn.commit(Nil, Truncate())
-      txn.commit(addA.remove :: addB :: Nil, Truncate())
+      winningTxn.commit(Nil, ManualUpdate)
+      txn.commit(addA.remove :: addB :: Nil, ManualUpdate)
       checkAnswer(log.update().allFiles.select("path"), Row("b") :: Nil)
     }
   }
 
+
+  val A_P1 = "part=1/a"
+  val B_P1 = "part=1/b"
+  val C_P1 = "part=1/c"
+  val C_P2 = "part=2/c"
+  val D_P2 = "part=2/d"
+  val E_P3 = "part=3/e"
+  val F_P3 = "part=3/f"
+  val G_P4 = "part=4/g"
+
+  private val addA_P1 = AddFile(A_P1, Map("part" -> "1"), 1, 1, dataChange = true)
+  private val addB_P1 = AddFile(B_P1, Map("part" -> "1"), 1, 1, dataChange = true)
+  private val addC_P1 = AddFile(C_P1, Map("part" -> "1"), 1, 1, dataChange = true)
+  private val addC_P2 = AddFile(C_P2, Map("part" -> "2"), 1, 1, dataChange = true)
+  private val addD_P2 = AddFile(D_P2, Map("part" -> "2"), 1, 1, dataChange = true)
+  private val addE_P3 = AddFile(E_P3, Map("part" -> "3"), 1, 1, dataChange = true)
+  private val addF_P3 = AddFile(F_P3, Map("part" -> "3"), 1, 1, dataChange = true)
+  private val addG_P4 = AddFile(G_P4, Map("part" -> "4"), 1, 1, dataChange = true)
+
+  test("allow concurrent commit on disjoint partitions") {
+    withLog(addA_P1 :: addD_P2 :: addE_P3 :: addD_P2.remove :: Nil) { log =>
+      val tx1 = log.startTransaction()
+      // TX1 reads P3 (but not P1)
+      val tx1Read = tx1.filterFiles(('part === 3).expr :: Nil)
+      assert(tx1Read.map(_.path) == E_P3 :: Nil)
+
+      val tx2 = log.startTransaction()
+      tx2.filterFiles()
+      // TX2 modifies only P1
+      tx2.commit(addB_P1 :: Nil, ManualUpdate)
+
+      // free to commit because P1 modified by TX2 was not read
+      tx1.commit(addC_P2 :: addE_P3.remove :: Nil, ManualUpdate)
+      checkAnswer(
+        log.update().allFiles.select("path"),
+        Row(A_P1) :: // start (E_P3 was removed by TX1)
+        Row(B_P1) :: // TX2
+        Row(C_P2) :: Nil) // TX1
+    }
+  }
+
+  test("allow concurrent commit on disjoint partitions reading all partitions") {
+    withLog(addA_P1 :: addD_P2 :: addD_P2.remove :: Nil) { log =>
+      val tx1 = log.startTransaction()
+      // TX1 read P1
+      tx1.filterFiles(('part isin 1).expr :: Nil)
+
+      val tx2 = log.startTransaction()
+      tx2.filterFiles()
+      tx2.commit(addC_P2 :: addD_P2.remove :: Nil, ManualUpdate)
```

Comment:
this file `addD_P2` already removed by initial commit, can we remove the same file multi-time?

Reply:
That's definitely a mistake. `addD_P2.remove` should not be in the initial commit of the log. In terms of guarantees, technically it is okay for a committed log to have duplicate removes. However, it is not okay for concurrent txns to concurrently attempt to remove the same file multiple times, because it breaks the semantic integrity of the second operation.

Reply:
True, please also change TX1's read to all partitions: `tx1.filterFiles(('part isin (1, 2)).expr :: Nil)`

Reply:
I am not sure I get why? The point of this test is to test that tx1 will be allowed because it is reading only part1 and tx2 is making changes to partition 2 only. If we change tx1 to read both partitions 1 and 2, then this tx1 is supposed to fail .. isnt it?

Reply:
You're right, however this is already checked in "allow concurrent commit on disjoint partitions". The idea behind this one was that TX1 reads all partitions the table has (almost like full table scan), and it is still allowed to commit. To make this correct, the TX2 should commit addG_P4 (not *_P2), so TX1 and TX2 would operate on disjoint partitions.

Reply:
Aah I get it now. That makes sense! Unfortunately, I merged this PR in our internal repo already to speed up things towards the 0.5.0 release. I will create a new internal PR and address this. 

Thank you very much for reviewing and working with us on this :)